duckduckgo/duckduckgo-autofill
# Refactor in-context state to feature file

**Reviewer:** 
**Asana:** 

## Description

@alistairjcbrown This is my 100% my issue for not seeing this earlier, but I was debugging some things recently I noticed that the purpose of `Settings` has not been communicated correctly in the past, something I will address.

For now, the short version is this: can we just create a feature-file for this work instead, which will allow you to hold your own internal state and not need to interact with settings as much (only for the true purpose, which was feature flags etc)

Please read through the comments and let me know if anything isn't clear, but the short version is this:

- `Settings` was designed as a cross-platform way to retrieve global settings/feature toggles etc, we shouldn't expand it's use to feature-related state
- When we do add new features, we can use the composition + dependency injection techniques instead, it's the exact same code if you look, but now the feature is self contained. It also then gives you a little container (the class) in which you can have your own internal state without this feature touching too many parts of the codebase.
- If you look at what I've removed from `Settings` and `extension.transport.js` I'm sure you'll be happy 😎

```js
export class InContextSignup {
    // your own internal state, instead of expanding the scope of `Settings`
    permanentlyDismissed = false
    initiallyDismissed = false

    /**
     * Pass the device as a constructor param, and then do everything as before
     * @param {import("./DeviceInterface/InterfacePrototype").default} device
     */
    constructor(device) {
        this.device = device
    }

    async init() { 
       // etc 
    } 
}

// then, you can do the composition + dependency injection like this:
class ExtensionInterface extends InterfacePrototype {
    inContextSignup = new InContextSignup(this)
}
```

## Steps to test

Can you run this locally and see if it works? I just put it together as a POC (I've not changed any functionality of course) but it would be good to get this resolved before your feature is rolled out :)
